### PR TITLE
Bug 1364233 - Add setting to force a group to require MFA and restrict users in that group who have not enabled MFA

### PR DESCRIPTION
**Pull request by dylanwh**

### CODE REVIEW

File: Bugzilla/Auth.pm [hunk 1]
```diff
@@ -111,6 +111,26 @@ sub login {
         });
     }
 
+    my $mfa_group = 'admin';
+    my %mfa_allowed_page = map { $_ => 1 } qw(index.cgi post_bug.cgi process_bug.cgi userprefs.cgi);
```
Comment (floatingatoll) [inline Bugzilla/Auth.pm]: This is a lot of work to support and opens some interesting MFA-bypass opportunities through process_bug.cgi. I think you'd be better off taking BMO offline for maintenance during a scheduled window and shipping this as a breaking change, without any exceptions other than `userprefs.cgi?tab=mfa` only.

File: Bugzilla/Auth.pm [hunk 2]
```diff
@@ -111,6 +111,26 @@ sub login {
...
+        && !$user->mfa
+        && $mfa_group
+        && $user->in_group($mfa_group) )
+    {
+        my $cgi = Bugzilla->cgi;
+        my $template = Bugzilla->template;
+        my $script = $cgi->script_name;
+        $script =~ s{^/}{};
+
+        unless ($mfa_allowed_page{$script}) {
```
Comment (floatingatoll) [inline Bugzilla/Auth.pm]: Per above, encourage replacing with 'userprefs.cgi?tab=mfa only', all other requests refused. (Even index.cgi, because that'll just direct them to ?tab=mfa anyways, as intended.)

File: template/en/default/mfa/page_requires_mfa.html.tmpl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[% PROCESS "global/field-descs.none.tmpl" %]
+
+[% PROCESS global/header.html.tmpl
+   title  = "This page requires 2FA",
```
Comment (floatingatoll) [inline template/en/default/mfa/page_requires_mfa.html.tmpl]: This is slightly incorrect. "Your security group memberships require you to configure 2FA to continue using Bugzilla." or "Please configure 2FA to continue." is more accurate, assuming a hard-line stance anyways.

File: Bugzilla/Auth.pm [hunk 4]
```diff
@@ -111,6 +111,26 @@ sub login {
         });
     }
 
+    my $mfa_group = 'admin';
+    my %mfa_allowed_page = map { $_ => 1 } qw(index.cgi post_bug.cgi process_bug.cgi userprefs.cgi);
+    if (   !i_am_webservice()
+        && $user->id
+        && !$user->mfa
+        && $mfa_group
```
Comment (floatingatoll) [inline Bugzilla/Auth.pm]: Note that eventually $mfa_group will be an array.

File: template/en/default/mfa/page_requires_mfa.html.tmpl [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+   style_urls = ['skins/standard/page.css']         
+%]
+
+<p>Because you are a member of [% mfa_group FILTER html %] you must have 2FA enabled.
```
Comment (floatingatoll) [inline template/en/default/mfa/page_requires_mfa.html.tmpl]: See prior note about $mfa_group being an array someday.

File: Bugzilla/Auth.pm (hunk 4 above)
Comment (dylanwh) [inline Bugzilla/Auth.pm]: Not really needed,  can just add more groups to the mfa_required group

File: template/en/default/mfa/page_requires_mfa.html.tmpl (hunk 5 above)
Comment (globau) [inline template/en/default/mfa/page_requires_mfa.html.tmpl]: i agree with dylan that `mfa_group` needn't be an array because of group inheritance, but the message here should reflect the group the user is a direct member of.

ie. if there's a `mfa-required` group which `mozilla-employee` is a member of, and i'm a member of `mozilla-employee`, this message should say "because you are a member of mozilla-employee", not "because you are a member of mfa-required".

**Comment by floatingatoll** [comment]
Mostly my question is; I’m in 15 security groups, of which I intend at
least 3 to become MFA required the instant y’all permit us.

So when I login and the home page refuses me for “you’ll need MFA to
continue”, does it list all three, or .. one only? Which one?

**Comment by dylanwh** [comment]
This patch is quite different than the original. It no longer indicates what group you're a member of.
The primary objective of this at the moment is Mozilla employees, and for users that have @mozilla.com emails, they can't opt-out of being in that group.

It also comes with a grace period (default: 7 days) where an ugly warning is displayed on every page. After that all activity is redirected to the mfa page.

the mfa page has conditional copy that hopefully provides a sufficient amount of hand-holding so that people can enable 2fa successfully.

File: Bugzilla/Config/Auth.pm [hunk 6]
```diff
@@ -145,6 +145,18 @@ sub get_param_list {
             type    => 't',
             default => '',
         },
+
+        {
+            name => 'mfa_group',
+            type => 't',
+            default => '',
+        },
```
Comment (globau) [inline Bugzilla/Config/Auth.pm]: this should be a group select, not a text field (as per most of the fields in Bugzilla::Config::GroupSecurity).

File: Bugzilla/Config/Auth.pm [hunk 7]
```diff
@@ -145,6 +145,18 @@ sub get_param_list {
...
+
+        {
+            name => 'mfa_group_grace_period',
+            type => 't',
+            default => '7',
```
Comment (globau) [inline Bugzilla/Config/Auth.pm]: needs check_numeric

File: Bugzilla/User.pm [hunk 8]
```diff
@@ -2228,10 +2250,8 @@ our %names_to_events = (
     'resolution'              => EVT_OPENED_CLOSED,
     'keywords'                => EVT_KEYWORD,
     'cc'                      => EVT_CC,
-    'bug_severity'            => EVT_PROJ_MANAGEMENT,
     'priority'                => EVT_PROJ_MANAGEMENT,
     'bug_status'              => EVT_PROJ_MANAGEMENT,
-    'target_milestone'        => EVT_PROJ_MANAGEMENT,
```
Comment (globau) [inline Bugzilla/User.pm]: these changes don't appear to be related to this bug

File: Bugzilla.pm [hunk 9]
```diff
@@ -348,21 +348,52 @@ sub login {
...
+                print Bugzilla->cgi->redirect("userprefs.cgi?tab=mfa");
+                exit;
+            }
+        }
+        else {
+            my $dbh = Bugzilla->dbh_main;
+            my $date = $dbh->sql_date_math( 'NOW()', '+', '?', 'DAY' );
+            my $user_id = $authenticated_user->id;
+            my ($mfa_required_date ) = $dbh->selectrow_array( "SELECT $date", undef, $grace_period );
+            warn $mfa_required_date;
```
Comment (globau) [inline Bugzilla.pm]: debugging?

File: template/en/default/account/prefs/mfa.html.tmpl [hunk 10]
```diff
@@ -6,6 +6,11 @@
   # defined by the Mozilla Public License, v. 2.0.
   #%]
 
+[%
+  SET MFA_HOWTO = "https://wiki.mozilla.org/BMO/UserGuide/Two-Factor_Authentication"
+      MFA_VIDEO = "#todo"
```
Comment (globau) [inline template/en/default/account/prefs/mfa.html.tmpl]: TODO items need doing before landing :)

File: template/en/default/account/prefs/mfa.html.tmpl [hunk 11]
```diff
@@ -120,15 +125,31 @@
         <p>
           <img src="images/duo.png" id="duo-logo" width="32" height="32">
           Verification with Duo Security will be performed before your account is updated.
-        </p>
```
Comment (globau) [inline template/en/default/account/prefs/mfa.html.tmpl]: won't removing this `</p>` result in unbalanced elements?

File: template/en/default/global/header.html.tmpl [hunk 12]
```diff
@@ -234,6 +234,9 @@
   <body
         class="[% urlbase.replace('^https?://','').replace('/$','').replace('[-~@:/.]+','-') FILTER css_class_quote %]
                skin-[% user.settings.skin.value FILTER css_class_quote %]
+               [% IF Bugzilla.request_cache.mfa_warning %]
+               mfa-warning
```
Comment (globau) [inline template/en/default/global/header.html.tmpl]: nit: indent the block, or replace with a one-liner `[% "mfa-warning" IF Bugzilla.request_cache.mfa_warning %]`

File: template/en/default/global/header.html.tmpl [hunk 13]
```diff
@@ -252,6 +255,16 @@
       </td>
       <td>
         [% Hook.process("message") %]
+        [% IF Bugzilla.request_cache.mfa_warning AND user.mfa_required_date %]
+        <span id="mfa-warning">
+          Please <a href="userprefs.cgi?tab=mfa">enabled two-factor authentication</a>
```
Comment (globau) [inline template/en/default/global/header.html.tmpl]: s/enabled/enable/

File: template/en/default/global/header.html.tmpl [hunk 14]
```diff
@@ -252,6 +255,16 @@
       </td>
       <td>
         [% Hook.process("message") %]
+        [% IF Bugzilla.request_cache.mfa_warning AND user.mfa_required_date %]
+        <span id="mfa-warning">
```
Comment (globau) [inline template/en/default/global/header.html.tmpl]: this block needs indentation  too.

File: Bugzilla.pm [hunk 15]
```diff
@@ -348,21 +348,52 @@ sub login {
...
+        if ( $grace_period == 0 || $expired) {
+            if (!$on_mfa_page) {
+                print Bugzilla->cgi->redirect("userprefs.cgi?tab=mfa");
+                exit;
+            }
+        }
+        else {
+            my $dbh = Bugzilla->dbh_main;
+            my $date = $dbh->sql_date_math( 'NOW()', '+', '?', 'DAY' );
+            my $user_id = $authenticated_user->id;
```
Comment (globau) [inline Bugzilla.pm]: `$user_id` is unused

File: Bugzilla/Auth.pm [hunk 16]
```diff
@@ -111,6 +111,8 @@ sub login {
         });
     }
 
+
+
```
Comment (globau) [inline Bugzilla/Auth.pm]: this change isn't required

File: template/en/default/mfa/page_requires_mfa.html.tmpl (hunk 5 above)
Comment (globau) [reply template/en/default/mfa/page_requires_mfa.html.tmpl]: this hasn't been addressed - displaying the the `mfa_group` name here doesn't provide any information to the user.  change it to either list the groups that are direct members of `mfa_group`, or make the wording generic - "You are a member of a Bugzilla group that requires you to enable 2FA."

File: Bugzilla/User.pm (hunk 8 above)
Comment (dylanwh) [reply Bugzilla/User.pm]: woops, those were accidental.

**Comment by dylanwh** [comment]
page_requires_mfa.html.tmpl  is a dead file, I'll remove it in next patchset

File: Bugzilla.pm [hunk 17]
```diff
@@ -348,21 +348,52 @@ sub login {
...
+
+        # decide if the user needs a warning or to be blocked.
+        my $date         = $authenticated_user->mfa_required_date('UTC');
+        my $grace_period = Bugzilla->params->{mfa_group_grace_period};
+        my $expired      = defined $date && $date < DateTime->now;
+        my $on_mfa_page  = $cgi->script_name eq '/userprefs.cgi' && $cgi->param('tab') eq 'mfa';
+
+        Bugzilla->request_cache->{mfa_warning} = 1;
+        Bugzilla->request_cache->{mfa_expired} = $expired;
+        Bugzilla->request_cache->{on_mfa_page} = $on_mfa_page;
```
Comment (globau) [inline Bugzilla.pm]: looks like `{mfa_expired}` and `{on_mfa_page}` aren't used

Comment (dylanwh) [reply Bugzilla.pm]: A change of name, s/mfa_expired/mfa_grace_period_expired/

**Comment by dylanwh** [comment]
Except for that error.

**Comment by globau** [comment]
i still see some review points from the last review not addressed (eg. the accidental removal of EVT_PROJ_MANAGEMENT).  can you ensure you've pushed up your latest changes.

**Comment by dylanwh** [comment]
Github UI disagreed with local checkout -- despite hash matching. I've just rebased it as one commit.
